mempool/mempool.go

```diff
@@ -0,0 +1,357 @@
+		}
+	case *core.L1HandlerTransaction:
+		// todo: verification of the L1 handler nonce requires checking the
+		// message nonce on the L1 Core Contract.
+	}
+	return nil
+}
+
+// Pop returns the transaction with the highest priority from the in-memory pool
+func (p *Pool) Pop() (BroadcastedTransaction, error) {
+	p.txnList.mu.Lock()
+	defer p.txnList.mu.Unlock()
+
+	if p.txnList.head == nil {
+		return BroadcastedTransaction{}, errors.New("transaction pool is empty")
+	}
+
+	headNode := p.txnList.head
+	p.txnList.head = headNode.Next
+	if p.txnList.head == nil {
+		p.txnList.tail = nil
+	}
+	p.txnList.len--
+
+	return headNode.Txn, nil
+}
+
+// Remove removes a set of transactions from the pool
+// todo: should be called by the builder to remove txns from the db everytime a new block is stored.
+// todo: in the consensus+p2p world, the txns should also be removed from the in-memory pool.
+func (p *Pool) Remove(hash ...*felt.Felt) error {
+	return errors.New("not implemented")
+}
+
+// Len returns the number of transactions in the in-memory pool
+func (p *Pool) Len() int {
+	return p.txnList.len
+}
+
+// Len returns the number of transactions in the persistent pool
+func (p *Pool) LenDB() (int, error) {
+	p.wg.Add(1)
+	defer p.wg.Done()
+	txn, err := p.db.NewTransaction(false)
+	if err != nil {
+		return 0, err
+	}
+	lenDB, err := p.lenDB(txn)
+	if err != nil {
+		return 0, err
+	}
+	return lenDB, txn.Discard()
+}
+
+func (p *Pool) lenDB(txn db.Transaction) (int, error) {
+	var l int
+	err := txn.Get(db.MempoolLength.Key(), func(b []byte) error {
+		l = int(new(big.Int).SetBytes(b).Int64())
+		return nil
+	})
+
+	if err != nil && errors.Is(err, db.ErrKeyNotFound) {
+		return 0, nil
+	}
+	return l, err
+}
+
+func (p *Pool) updateLen(txn db.Transaction, l int) error {
+	return txn.Set(db.MempoolLength.Key(), new(big.Int).SetInt64(int64(l)).Bytes())
+}
+
+func (p *Pool) Wait() <-chan struct{} {
+	return p.txPushed
+}
+
+func (p *Pool) headHash(txn db.Transaction, head *felt.Felt) error {
+	return txn.Get(db.MempoolHead.Key(), func(b []byte) error {
+		head.SetBytes(b)
+		return nil
+	})
+}
+
+func (p *Pool) updateHead(txn db.Transaction, head *felt.Felt) error {
+	return txn.Set(db.MempoolHead.Key(), head.Marshal())
+}
+
+func (p *Pool) tailValue(txn db.Transaction, tail *felt.Felt) error {
+	return txn.Get(db.MempoolTail.Key(), func(b []byte) error {
+		tail.SetBytes(b)
+		return nil
+	})
+}
+
+func (p *Pool) updateTail(txn db.Transaction, tail *felt.Felt) error {
+	return txn.Set(db.MempoolTail.Key(), tail.Marshal())
+}
+
+func (p *Pool) dbElem(txn db.Transaction, itemKey *felt.Felt) (storageElem, error) {
+	var item storageElem
+	keyBytes := itemKey.Bytes()
+	err := txn.Get(db.MempoolNode.Key(keyBytes[:]), func(b []byte) error {
+		return encoder.Unmarshal(b, &item)
+	})
+	return item, err
+}
+
+func (p *Pool) putdbElem(txn db.Transaction, itemKey *felt.Felt, item *storageElem) error {
```

Comment:
And here `writedbPoolTxn(...)`. Also, I believe the `itemKey` field here is redundant since it can obtained from `item`.